lib/skate/detours/missed_stops.ex

```diff
@@ -0,0 +1,177 @@
+          [stop]
+      end
+
+      []
+    else
+      Enum.take(remaining_segments, end_count)
+    end
+  end
+
+  @spec segment_shape_by_stops(
+          shape :: [Util.Location.From.t()],
+          stops :: [Util.Location.From.t()]
+        ) :: [Skate.Detours.ShapeSegment.t()]
+  defp segment_shape_by_stops([] = _shape, [] = _stops),
+    do: []
+
+  defp segment_shape_by_stops(shape, [] = _stops),
+    # If there are no stops, return the shape
+    do: [%Skate.Detours.ShapeSegment{points: shape, stop: :none}]
+
+  defp segment_shape_by_stops(shape, stops) do
+    # Find the stop closest to the shape
+    %{index: anchor_stop_index, shape_dist: %{index: shape_point_anchor_index}} =
+      find_stop_closest_to_shape(shape, stops)
+
+    # Split the shape and stops at the anchor point
+    # Add one to the Shape Point Anchor Index so that the point ends up in `left_shape`
+    {left_shape, right_shape} = Enum.split(shape, shape_point_anchor_index + 1)
+
+    {left_stops, [anchor | right_stops]} = Enum.split(stops, anchor_stop_index)
+
+    # Process `left_shape` into segments
+    {left_segments, [anchor_segment]} =
+      Enum.split(segment_shape_by_stops(left_shape, left_stops), -1)
+
+    # Take last segment, which should not have a stop,
+    # and reconnect it with the anchor stop
+    anchor_segment = %Skate.Detours.ShapeSegment{
+      (%Skate.Detours.ShapeSegment{stop: :none} = anchor_segment)
+      | stop: anchor
+    }
+
+    left_segments ++ [anchor_segment] ++ segment_shape_by_stops(right_shape, right_stops)
+  end
+
+  @spec find_stop_closest_to_shape(
+          shape :: [Util.Location.From.t()],
+          stops :: [Util.Location.From.t()]
+        ) ::
+          %{
+            stop: Util.Location.From.t(),
+            index: non_neg_integer(),
+            shape_dist: point_dist_index()
+          }
+
+  defp find_stop_closest_to_shape(shape, stops) do
+    stops
+    |> Enum.with_index(fn stop, index ->
+      %{
+        stop: stop,
+        index: index,
+        shape_dist: closest_point(shape, stop)
+      }
+    end)
+    |> Enum.min_by(fn %{shape_dist: %{dist: dist}} -> dist end)
+  end
+
+  @spec get_index_by_min_dist(
+          shape_segments :: [Skate.Detours.ShapeSegment.t()],
+          reference :: Util.Location.From.t()
+        ) ::
+          %{
+            segment: Skate.Detours.ShapeSegment.t(),
+            sorted_points: [Util.Location.From.t()],
+            index: non_neg_integer()
+          }
+  defp get_index_by_min_dist(shape_segments, reference) do
+    shape_segments
+    |> Enum.with_index(fn %Skate.Detours.ShapeSegment{points: points} = segment, idx ->
+      points = closest_point(points, reference)
+      # Pick closest points within fuzzy_dist=(0)
+      # Discard adjacent points
+      # Pick highest rank or return :indeterminate
+      %{segment: segment, sorted_points: points, index: idx}
+    end)
+    |> Enum.min_by(fn %{sorted_points: %{dist: dist}} -> dist end)
+  end
```

Comment:
```suggestion
  @spec get_index_by_min_dist(
          shape_segments :: [Skate.Detours.ShapeSegment.t()],
          reference :: Util.Location.From.t()
        ) ::
          %{
            segment: Skate.Detours.ShapeSegment.t(),
            closest_point: point_dist_index(),
            index: non_neg_integer()
          }
  defp get_index_by_min_dist(shape_segments, reference) do
    shape_segments
    |> Enum.with_index(fn %Skate.Detours.ShapeSegment{points: points} = segment, idx ->
      closest_point = closest_point(points, reference)
      # Pick closest points within fuzzy_dist=(0)
      # Discard adjacent points
      # Pick highest rank or return :indeterminate
      %{segment: segment, closest_point: closest_point, index: idx}
    end)
    |> Enum.min_by(fn %{sorted_points: %{dist: dist}} -> dist end)
  end
```

Nit: Both the names `points` and `sorted_points` are a bit confusing now that each refers to a singular point.

(Also I know that Elixir lets you re-bind variable names to new data, but I prefer to avoid it when possible, since then it gets hard to tell which definition is the most recent)

I'm also wondering why Dialyxir didn't complain about the type of `sorted_points` being wrong. (I think it was wrong?)

Reply:
0146770708ef7a8b020c0b62544d8bd38dc25fad